src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/opt.cpp

```diff
@@ -1653,6 +1672,63 @@ HostGatherDQ::HostGatherDQ(Context::Ref ctx) {
     register_matcher(std::make_shared<opp::Matcher>(qmul, "HostGatherDQ"), std::move(callback));
 }
 
+HostGatherNF4::HostGatherNF4(Context::Ref ctx) {
+    auto pids = opp::wrap_type<ov::op::v0::Parameter>();
+    auto cvtids = opp::wrap_type<ov::op::v0::Convert>({pids});
+
+    auto qweight = opp::wrap_type<ov::op::v0::Parameter>();
+    auto qcvtw = opp::wrap_type<ov::op::v0::Convert>({qweight});
+    auto qweightt = opp::wrap_type<ov::op::v0::Constant>();
+    auto qcvtwt = opp::wrap_type<ov::op::v0::Convert>({qweightt});
+    auto qweightg = opp::wrap_type<ov::op::v8::Gather>({qcvtwt, qcvtw, opp::any_input()});
+
+    auto qcoeff = opp::wrap_type<ov::op::v0::Parameter>();
+
+    auto qmul = opp::wrap_type<ov::op::v1::Multiply>({qweightg, qcoeff});
+    auto qcvtmul = opp::wrap_type<ov::op::v0::Convert>({qmul});
+
+    auto qgthr = opp::wrap_type<ov::op::v8::Gather>({qcvtmul, cvtids, opp::any_input()});
+
+    auto callback = [=](ov::pass::pattern::Matcher& m) {
+        auto& node_to_output = m.get_pattern_value_map();
+        const auto& matched_out_mul = node_to_output.at(qmul);
+        auto out_shape = matched_out_mul.get_shape();
+
+        if (out_shape.size() == 2 && out_shape.back() >= 2048 &&
+            node_to_output.at(qweight).get_element_type() == ov::element::u4) {
+            auto matched_node_qweight = node_to_output.at(qweight).get_node_shared_ptr();
+            auto matched_node_qcoeff = node_to_output.at(qcoeff).get_node_shared_ptr();
+            auto matched_node_ids = node_to_output.at(pids).get_node_shared_ptr();
+            auto matched_node_qgthr = node_to_output.at(qgthr).get_node_shared_ptr();
+            auto matched_node_qweightt = node_to_output.at(qweightt).get_node_shared_ptr();
+
+            auto matched_qweight = std::static_pointer_cast<ov::op::v0::Parameter>(matched_node_qweight);
+            auto matched_qcoeff = std::static_pointer_cast<ov::op::v0::Parameter>(matched_node_qcoeff);
+            auto matched_ids = std::static_pointer_cast<ov::op::v0::Parameter>(matched_node_ids);
+            auto matched_qweightt = std::static_pointer_cast<ov::op::v0::Constant>(matched_node_qweightt);
+
+            // Need to gather the weight into f16 first
+            auto fp16weight = ctx.get().gather_nf4(matched_qweight,
+                                                   ov::npuw::util::copy_tensor_from_const(matched_qweightt),
+                                                   ov::element::f16,
+                                                   matched_qweightt->get_data_ptr());
```

Comment:
so.. what's the reason to make copy right here? why don't you pass the tensor itself like it was supposed to be done?

if you just pass the tensor, you don't need to pass this pointer everywhere.

Reply:
Removed, using tensor data instead